File: src/modules/cmdpal/extensionsdk/Microsoft.CommandPalette.Extensions.Toolkit/Separator.cs
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using Windows.Foundation.Collections;

namespace Microsoft.CommandPalette.Extensions.Toolkit;

public partial class Separator : BaseObservable, IListItem, ISeparatorContextItem, ISeparatorFilterItem, IExtendedAttributesProvider
{
    private readonly PropertySet _extendedAttributes = new();
    private ICommand? _sectionCommand;

    public IDetails? Details => null;

    public string? Section { get; private set; }

    public ITag[]? Tags => null;

    public string? TextToSuggest => null;

    public ICommand? Command => null;

    /// <summary>
    /// Gets or sets the optional command displayed alongside this separator when it is used as a section header.
    /// </summary>
    /// <remarks>
    /// <see cref="Command"/> remains <see langword="null"/> so hosts continue to identify this object as a separator.
    /// The section command is exposed to the host through <see cref="IExtendedAttributesProvider"/>.
    /// </remarks>
    public ICommand? SectionCommand
    {
        get => _sectionCommand;
        set
        {
            if (EqualityComparer<ICommand?>.Default.Equals(value, _sectionCommand))
            {
                return;
            }

            _sectionCommand = value;
            if (value is null)
            {
                _extendedAttributes.Remove(WellKnownExtensionAttributes.SectionCommand);
            }
            else
            {
                _extendedAttributes[WellKnownExtensionAttributes.SectionCommand] = value;
            }

            OnPropertyChanged();
        }
    }

    public IIconInfo? Icon => null;

    public IContextItem[]? MoreCommands => null;

    public string? Subtitle => null;

    public string? Title
    {
        get => Section;
        set
        {
            if (Section != value)
            {
                Section = value;
                OnPropertyChanged();
                OnPropertyChanged(nameof(Section));
            }
        }
    }

    public Separator(string? title = "")
    {
        Section = title ?? string.Empty;
    }

    public Separator(string? title, ICommand? sectionCommand)
        : this(title)
    {
        SectionCommand = sectionCommand;
    }

    public IDictionary<string, object> GetProperties() => _extendedAttributes;
}

File: src/modules/cmdpal/ts-sdk/src/base/Separator.ts
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

import type { ICommand, IListItem } from '../types.js';

/**
 * A visual divider for list pages. With no title it renders as a horizontal
 * line; with a title it renders as a section header.
 *
 * @example
 * ```typescript
 * new Separator();              // horizontal line
 * new Separator('Recent');      // section header
 * ```
 */
export class Separator implements IListItem {
  /** Placeholder command; separators are not invokable. */
  readonly command: ICommand;
  /** Title text; empty for a plain divider. */
  title: string;
  /** Section header text when the separator names a group. */
  section?: string;

  /** Marker used by the runtime to serialize this item as a separator. */
  readonly isSeparator = true;

  /**
   * Creates a separator.
   *
   * @param title Optional section header text. When omitted the separator
   * renders as a plain horizontal line.
   */
  constructor(title = '') {
    this.title = title;
    this.section = title || undefined;
    this.command = { id: `separator:${title}`, name: '' };
  }
}
